xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -89,29 +88,66 @@ pmap_registry_get(xlator_t *this)
     return pmap;
 }
 
-/*
- * The "destroy" argument avoids a double search in pmap_registry_remove - one
- * to find the entry in the table, and the other to find the particular
- * brickname within that entry (which might cover multiple bricks).  We do the
- * actual deletion here by "whiting out" the brick name with spaces.  It's up
- * to pmap_registry_remove to figure out what to do from there.
- */
+/* Randomly selecting a port number between base_port and max_port,
+   then checking if the port is free or not using pmap_port_isfree().
+   If the port is free, return the port number
+*/
+
 int
-pmap_registry_search(xlator_t *this, const char *brickname,
-                     gf_pmap_port_type_t type, gf_boolean_t destroy)
+pmap_port_alloc(xlator_t *this)
 {
     struct pmap_registry *pmap = NULL;
     int p = 0;
+
+    GF_ASSERT(this);
+
+    pmap = pmap_registry_get(this);
+
+    while (true) {
+        p = (rand() % (pmap->max_port - pmap->base_port + 1)) + pmap->base_port;
+        if (pmap_port_isfree(p)) {
+            break;
+        }
+    }
+
+    return p;
+}
+
+/* pmap_assign_port does a pmap_registry_remove followed by pmap_port_alloc,
+ * the reason for the former is to ensure we don't end up with stale ports
+ */
+int
+pmap_assign_port(xlator_t *this, int old_port, char *path)
```

Comment:
What prevents two concurrent calls to `pmap_assign_port()` to return the same port ?

Reply:
I think glusterd takes care of that, the reason being we have the same mechanism being used currently and we don't face any such race condition.

Reply:
While I agree that the old implementation didn't use locks either, it had a table where assigned ports were marked so that other calls to assign a port don't return the same port. In this implementation we are not checking if a port has already been assigned.

So, if the old implementation was wrong, this new version is worse because the race window is bigger.

We need to be sure that these functions are called in a safe way or we'll need to implement some additional locking/checks here.

Reply:
To me, the whole idea is to randomize. We'll then use that port. We can anyway fail using that port, because another application may have opted using it, so we'll have to go back and ask for a new, again randomized, port.

Reply:
The old implementation wasn't wrong, rather it was using stale entries so instead of checking the ports at the startup and deciding free/in-use we are checking it only while using it.

Here, the function `pmap_port_isfree` checks for the free port and returns accordingly.....so if a brick is assigned a port number and is used by some other process before the brick process actually starts on that port then a new port is assigned to brick and it is started on the new port.

Reply:
> To me, the whole idea is to randomize. We'll then use that port. We can anyway fail using that port, because another application may have opted using it, so we'll have to go back and ask for a new, again randomized, port.

The problem is that I'm not sure if we do that. Looking at the code, a brick can be started synchronously or asynchronously. If it's started synchronously, we check for EADDRINUSE error and retry. However if it's started asynchronously, I don't see where we do that check (but maybe I've missed something).

After a quick look at the code, it seems we only start bricks synchronously after an add brick operation. Other operations seem to start them asynchronously, so probably not checking if the address already in use.

I've only done a quick look at the code, so maybe I'm wrong.

Reply:
> The old implementation wasn't wrong, rather it was using stale entries so instead of checking the ports at the startup and deciding free/in-use we are checking it only while using it.
> 
> Here, the function `pmap_port_isfree` checks for the free port and returns accordingly.....so if a brick is assigned a port number and is used by some other process before the brick process actually starts on that port then a new port is assigned to brick and it is started on the new port.

The difference is that in the previous implementation, a port just assigned to be used by a brick (besides the ports used during the initial scan) was also marked as busy, while we are not doing it now.

Anyway, as you say, as soon as a port is returned, it could become in use by another application before the brick starts, so we still need to handle this case correctly, and if we do so, it doesn't matter here.

Reply:
>Anyway, as you say, as soon as a port is returned, it could become in use by another application before the brick starts, so we still need to handle this case correctly, and if we do so, it doesn't matter here.
 
Yeah, a check for that is present in the caller function which takes care of that part.https://bugzilla.redhat.com/show_bug.cgi?id=1865796 like something similar happened in this bug but we have a check to prevent this scenario.

Reply:
For starting the bricks we have the same code being used at all places which is by calling this function `glusterd_brick_start` which will handle the EADDRINUSE error and restart the bricks on new ports if we get the error.

Reply:
Yes, `glusterd_brick_start` is the common function to start the bricks, but it has an argument called `wait`. If `wait` is true, the brick is executed synchronously and the return code is checked, but if it's false, `runner_run_nowait()` is called and it doesn't seem we check if the program succeeded or not to retry with another port in case of error.

The only functions I see that pass `wait == true` are `glusterd_start_volume()` and `glusterd_op_perform_add_bricks()`, but there are several operations that use `wait == false`, so those operations can fail to start a brick if the address is in use (for example it seems that if glusterd tries to launch bricks for an already started volume, it will use `wait == false`. Snapshots also seem to have the same problem).